flow/project.py

```diff
@@ -2312,11 +2329,37 @@ def hello(job):
         cls._OPERATION_FUNCTIONS.append((name, func))
         return func
 
+    @classmethod
+    def _collect_operations(cls):
+        "Collect all operations that were add via decorator."
+        operations = []
+        for cls_ in cls.__mro__:
+            operations.extend(getattr(cls_, '_OPERATION_FUNCTIONS', []))
+        return operations
+
+    @classmethod
+    def _collect_pre_conditions(cls):
+        "Collect all pre-conditions that were add via decorator."
+        ret = defaultdict(list)
+        for cls_ in cls.__mro__:
+            for func, conds in getattr(cls_, '_OPERATION_PRE_CONDITIONS', dict()).items():
+                ret[func].extend(conds)
+        return ret
+
+    @classmethod
+    def _collect_post_conditions(cls):
+        "Collect all post-conditions that were add via decorator."
+        ret = defaultdict(list)
+        for cls_ in cls.__mro__:
+            for func, conds in getattr(cls_, '_OPERATION_POST_CONDITIONS', dict()).items():
+                ret[func].extend(conds)
+        return ret
```

Comment:
While I don't normally advocate for deduping code when there's just two copies (our favorite rule of 3!), in this case I think it would be worthwhile to combine the two collection functions for conditions. The code is identical except for the attribute, and I think it's worth consolidating any mro logic as much as possible to make it easier for devs/contributors who are less familiar with python internals.

Reply:
I split them, because I thought it would be good to be able to access these two different attributes separately. It comes with code duplication, but like you said, only *once*, so not that big of a problem. I would advocate against deduping, but I don't feel strongly about it.

Reply:
I agree that being able to access them separately is good. An alternative would be to define a third (helper) function that contains all of the logic and takes a `key` or `attr` argument, then having either of these functions be one-liners that call the helper function with the appropriate attribute to get. To reiterate, my primary reasoning is that centralizing mro logic makes it easier for future developers who might not know how that works as well. If you prefer not to do this, then we can just leave it as is.